tests/test_runner.py

```diff
@@ -0,0 +1,286 @@
+"""
+Copyright 2018, Oath Inc.
+Licensed under the terms of the Apache 2.0 license. See LICENSE file in project root for terms.
+"""
+import os
+import unittest
+
+from mock import patch, MagicMock
+
+from yahoo_panoptes.framework.plugins.panoptes_base_plugin import PanoptesPluginInfo, PanoptesBasePlugin
+from yahoo_panoptes.polling.polling_plugin import PanoptesPollingPlugin
+from yahoo_panoptes.framework.resources import PanoptesContext, PanoptesResource
+from yahoo_panoptes.framework.plugins.runner import PanoptesPluginRunner, PanoptesPluginWithEnrichmentRunner
+from yahoo_panoptes.framework.metrics import PanoptesMetric, PanoptesMetricsGroupSet
+
+from .test_framework import PanoptesTestKeyValueStore, panoptes_mock_kazoo_client, panoptes_mock_redis_strict_client
+
+
+_TIMESTAMP = 1
+
+
+def _callback(*args):
+    pass
+
+
+def _callback_no_args():
+    pass
+
+
+class PanoptesTestPlugin(PanoptesBasePlugin):
+    name = None
+    signature = None
+    data = {}
+
+    execute_now = True
+    plugin_object = None
+
+    def run(self, context):
+        pass
+
+
+class PanoptesTestPluginRaiseException:
+    name = None
+    version = None
+    last_executed = None
+    last_executed_age = None
+    last_results = None
+    last_results_age = None
+    moduleMtime = None
+    configMtime = None
+
+    signature = None
+    data = {}
+
+    execute_now = True
+    lock = "dummy"
+
+    def run(self, context):
+        raise Exception
+
+
+class MockPluginExecuteNow:
+    execute_now = False
+
+
+class MockPluginLockException:
+    name = None
+    signature = None
+    data = {}
+
+    execute_now = True
+    lock = MagicMock(side_effect=Exception)
+
+
+class MockPluginLockNone:
+    name = None
+    signature = None
+    data = {}
+
+    execute_now = True
+    lock = None
+
+
+class TestPanoptesPluginRunner(unittest.TestCase):
+    @patch('redis.StrictRedis', panoptes_mock_redis_strict_client)
+    @patch('kazoo.client.KazooClient', panoptes_mock_kazoo_client)
+    def setUp(self):
+        self.my_dir, self.panoptes_test_conf_file = _get_test_conf_file()
+        self._panoptes_context = PanoptesContext(self.panoptes_test_conf_file,
+                                                 key_value_store_class_list=[PanoptesTestKeyValueStore],
+                                                 create_message_producer=False, async_message_producer=False,
+                                                 create_zookeeper_client=True)
+        self._runner_class = PanoptesPluginRunner
+
+    def test_basic_operations(self):
+        runner = self._runner_class("Test Polling Plugin", "polling", PanoptesPollingPlugin, PanoptesPluginInfo,
+                                    None, self._panoptes_context, PanoptesTestKeyValueStore,
+                                    PanoptesTestKeyValueStore, PanoptesTestKeyValueStore, "plugin_logger",
+                                    PanoptesMetricsGroupSet, _callback)
+        runner.execute_plugin()
```

Comment:
Since the plugin runner code is written to 'eat' errors and just log them instead of raising exceptions, the right thing to do would be to test logger output for each test. Note that this is not because we want unit testing for logging, but so that we can actually assert that what we expected to happen actually happened.

Reply:
Makes sense. QQ: Would something like this suffice?
`    def test_basic_operations(self):
        runner = self._runner_class("Test Polling Plugin", "polling", PanoptesPollingPlugin, PanoptesPluginInfo,
                                    None, self._panoptes_context, PanoptesTestKeyValueStore,
                                    PanoptesTestKeyValueStore, PanoptesTestKeyValueStore, "plugin_logger",
                                    PanoptesMetricsGroupSet, _callback)
        mock_info = MagicMock()
        with patch('yahoo_panoptes.framework.plugins.runner.PanoptesPluginRunner.info', mock_info):
            runner.execute_plugin()
            assert mock_info.call_count == 7 `